iscservice/demo/util.c

```diff
@@ -0,0 +1,710 @@
+{
+	m0_spiel_rconfc_stop(spiel);
+	m0_spiel_fini(spiel);
+}
+
+static bool conf_obj_is_svc(const struct m0_conf_obj *obj)
+{
+	return m0_conf_obj_type(obj) == &M0_CONF_SERVICE_TYPE;
+}
+
+struct m0_rpc_link * m0util_svc_rpc_link_get(struct m0_fid *svc_fid)
+{
+	struct m0_reqh *reqh = &m0_instance->m0c_reqh;
+	struct m0_reqh_service_ctx *ctx;
+	struct m0_pools_common *pc = reqh->rh_pools;
+
+	m0_tl_for(pools_common_svc_ctx, &pc->pc_svc_ctxs, ctx) {
+		if (m0_fid_eq(&ctx->sc_fid, svc_fid))
+			return &ctx->sc_rlink;
+	} m0_tl_endfor;
+
+	return NULL;
+}
+
+/*
+ * Loads a library into m0d instances.
+ */
+int m0util_isc_api_register(const char *libpath)
+{
+	int                     rc;
+	struct m0_reqh         *reqh = &m0_instance->m0c_reqh;
+	struct m0_confc        *confc;
+	struct m0_conf_root    *root;
+	struct m0_conf_process *proc;
+	struct m0_conf_service *svc;
+	struct m0_conf_diter    it;
+
+	rc = spiel_prepare(&spiel_inst);
+	if (rc != 0) {
+		fprintf(stderr, "error! spiel initialization failed");
+		return rc;
+	}
+
+	confc = m0_reqh2confc(reqh);
+	rc = m0_confc_root_open(confc, &root);
+	if (rc != 0) {
+		m0util_spiel_destroy(&spiel_inst);
+		return rc;
+	}
+
+	rc = m0_conf_diter_init(&it, confc,
+				&root->rt_obj,
+				M0_CONF_ROOT_NODES_FID,
+				M0_CONF_NODE_PROCESSES_FID,
+				M0_CONF_PROCESS_SERVICES_FID);
+	if (rc != 0) {
+		m0_confc_close(&root->rt_obj);
+		m0util_spiel_destroy(&spiel_inst);
+		return rc;
+	}
+
+	while (M0_CONF_DIRNEXT ==
+	       (rc = m0_conf_diter_next_sync(&it, conf_obj_is_svc))) {
+
+		svc = M0_CONF_CAST(m0_conf_diter_result(&it), m0_conf_service);
+		if (svc->cs_type != M0_CST_ISCS)
+			continue;
+		proc = M0_CONF_CAST(m0_conf_obj_grandparent(&svc->cs_obj),
+				    m0_conf_process);
+		rc = m0_spiel_process_lib_load(&spiel_inst, &proc->pc_obj.co_id,
+					       libpath);
+		if (rc != 0) {
+			fprintf(stderr, "error! loading the library %s failed "
+				        "for process "FID_F": rc=%d\n", libpath,
+					FID_P(&proc->pc_obj.co_id), rc);
+			m0_conf_diter_fini(&it);
+			m0_confc_close(&root->rt_obj);
+			m0util_spiel_destroy(&spiel_inst);
+			return rc;
+		}
+	}
+	m0_conf_diter_fini(&it);
+	m0_confc_close(&root->rt_obj);
+	m0util_spiel_destroy(&spiel_inst);
+
+	return 0;
+}
+
+int m0util_isc_nxt_svc_get(struct m0_fid *svc_fid, struct m0_fid *nxt_fid,
+			   enum m0_conf_service_type s_type)
+{
+	struct m0_reqh_service_ctx *ctx;
+	struct m0_reqh             *reqh = &m0_instance->m0c_reqh;
+	struct m0_pools_common     *pc = reqh->rh_pools;
+	struct m0_fid               current_fid = *svc_fid;
+
+	m0_tl_for(pools_common_svc_ctx, &pc->pc_svc_ctxs, ctx) {
+		if (ctx->sc_type == s_type) {
+			if (m0_fid_eq(&current_fid, &M0_FID0)) {
+				*nxt_fid = ctx->sc_fid;
+				return 0;
+			} else if (m0_fid_eq(svc_fid, &ctx->sc_fid))
+				current_fid = M0_FID0;
+		}
+	} m0_tl_endfor;
+
+	*nxt_fid = M0_FID0;
+	return -ENOENT;
+}
+
+int m0util_isc_req_prepare(struct m0util_isc_req *req, struct m0_buf *args,
```

Comment:
Also, this function should be part of Motr, as a public API.
So, isc applications can use it directly.

Reply:
In which case, we should probably move `isc_req_replied()` to libmotr as well, along with the global `isc_sem` variable. I'm not sure @nikitadanilov will be happy about that.